src/shared/dissect-image.c

```diff
@@ -2004,10 +2004,39 @@ static int mount_partition(
 
         if (p) {
                 if (m->fsmount_fd >= 0) {
-                        /* Case #1: Attach existing fsmount fd to the file system */
-
-                        if (move_mount(m->fsmount_fd, "", -EBADF, p, MOVE_MOUNT_F_EMPTY_PATH) < 0)
-                                return -errno;
+                        bool mount_beneath = FLAGS_SET(flags, DISSECT_IMAGE_TRY_ATOMIC_MOUNT_EXCHANGE);
+
+                        /* Case #1: Attach existing fsmount fd to the file system. First, try to mount
+                         * beneath an existing mount point, and if that works, umount the old mount, which
+                         * is now at the top. This will ensure we can atomically replace a mount. Note that
+                         * this works also in the case where there are submounts down the tree, as long as
+                         * there's no propagation. If this is not supported (minimum kernel v6.5), or if
```

Comment:
So propagation is allowed. It's just that specific propagation layouts are forbidden. Specifically:
(1) The parent mount of the target mount (the one you're trying to mount beneath) propagates to the target mount.
(2) The parent mount of the target mount (the one you're trying to mount beneath) propagates to the source mount.

In both those cases you would either end up with a copy of the source mount (the fsmount() fd that you're trying to mount beneath) on top - case (1) - or the source mount itself on top - case (2). Which makes this whole exercise pointless. So these two propagation layouts are forbidden. Otherwise propagation is fine.